### DISCUSSION TITLE
Getting all possible states

### DISCUSSION THREAD

**Discussion by n-sviridenko** [statelyai/xstate#3041] [Q&A]
How can we know all possible states the machine can transition to from its current state?

E.g. we have `pending -> started -> finished` & `pending -> started -> failed` machine; We want to have a function which will accept `started` and return `[finished, failed]`

**Comment by davidkpiano** [comment]
Given a `State` object, you can iterate through `state.nextEvents`:

```js
const getNextStates = (state) => {
  return state.nextEvents.map(nextEvent => {
    return machine.transition(state, nextEvent);
  })
}
```

This will work for simple events (e.g., `{ type: 'someEvent' }`, but for events that take payload, you need to be a little more creative. I would do something like:

```js
const eventPayloads = {
  greet: [{ message: 'hello' }, { message: 'goodbye' }],
  // ...
}

const getNextStates = (state) => {
  return state.nextEvents.flatMap(nextEvent => {
    const payloads = eventPayloads[nextEvent];

    if (payloads) {
      return payloads.map(payload => machine.transition(state, { type: nextEvent, ...payload });
    }
    return machine.transition(state, nextEvent);
  }
}
```

**Comment by n-sviridenko** [comment]
We need to do this in a list (we use it as a CRM) with 1000 rows (clients). Each of them has a dropdown (status) with all possible values. These options have to be filtered this way . How can we make it performant with React?

**Reply by davidkpiano** [reply]
Look into cache/memoization patterns, e.g. with `useMemo(...)`: https://reactjs.org/docs/hooks-reference.html#usememo
